WIP: Documentation Guides (#531)
mosaicml/composer: docs/source/trainer/using_the_trainer.rst

Using the Trainer
=================

The Composer :class:`.Trainer` implements a highly-optimized PyTorch training loop for neural networks. Using the trainer gives you several superpowers:

-  Easily insert our library of efficiency speedups methods into the
   trainer loop and compose them to train more efficiently and build
   better models.
-  Strong optimized baseline implementations to kick off your deep
   learning work, with reproducible results in time-to-train and
   accuracy.
-  Integrations with your favorite model hubs:
   🤗 `Transformers`_, `TIMM`_, and `torchvision`_.
-  Iterate faster! We take care of performance and efficiency.

.. note::

    We use the two-way callback system developed by (`Howard et al,
    2020 <https://arxiv.org/abs/2002.04688>`__) to flexible add the logic of
    our speed-up methods during training.

Below are simple examples for getting started with the Composer Trainer
along with code snippets for more advanced usage such as using speedup
methods, checkpointing, distributed training.

Getting Started
---------------

Create a model class that meets the :class:`.ComposerModel` interface,
minimally implementing the following methods:

> -  ``def forward(batch) -> outputs`` : computes the forward pass based

Author: wouldnt this be brittle if the line numbers change in the code?

> -  ``def forward(batch) -> outputs`` : computes the forward pass based

Reviewer: It is possible to also select lines between a start-string and end-string instead of line numbers

A few tips and tricks for using our Trainer:

-  For time-related inputs, such as the ``max_duration`` above, we
   support both an integer (which we assume is epochs), or as a string.
   For example, ``"10ba"`` means 10 minibatches or steps, and ``"10ep"``
   denotes 10 epochs.
-  If you are using gradient accumulation, the ``batch_size`` in your
   dataloaders should be the macrobatch size — the batch size of your
   optimization update. For example, with ``grad_accum=2`` and
   ``batch_size=2048`` , the train runs through two microbatches of 1024
   each, then performs a gradient update step.
-  At any time, most of the relevant quantities for debugging are
   centralized into one variable: :class:`.State`.
-  We have an easy abstraction of tracking :class:`.Time`, see the Time guide.

For a full list of Trainer options, see :class:`.Trainer`. Below, we
illustrate some example use cases.

Training Loop
~~~~~~~~~~~~~

Behind the scenes, our trainer handles much of the engineering for
distributed training, gradient accumulation, device movement, gradient
scaling, and others. The *pseudocode* for our trainer loop as it
interacts with the :class:`.ComposerModel` is as follows:

> .. code:: python

Reviewer: I feel like this pseudocode is too simple -- it doesn't show events; is missing schedulers; gradient accumulation, etc...
I do like the pseudocode in the event docstring. Maybe we can get to the event section in this guide, and then include the pseudocode there.

Author: This is primarily to explain how our trainer loop interacts with the `ComposerModel` methods, and not everything else. Is that OK then?

   # training loop
   for batch in train_dataloader:

       outputs = model.forward(batch)
       loss = model.loss(outputs, batch)

       loss.backward()
       optimizer.step()

   # eval loop
   for batch in eval_dataloader:
       outputs, targets = model.validate(batch)
       metrics.update(outputs, target)

> For the actual code, see :meth:`.Trainer._train_batch_inner` and the :meth:`.Trainer.eval` methods.

Reviewer: Private methods won't autolink since they are not documented.

```suggestion
For the actual code, see :meth:`.Trainer.fit` and the :meth:`.Trainer.eval` methods.
```

Quick Tour
----------

Below is a quick tour of various elements with code snippets for your reference. See the more
detailed sections in the navigation menu.

Events & State
~~~~~~~~~~~~~~

The core principle of the Composer trainer is to make it easy to inject
custom logic to run at various points in the training loop. To do this,
we have events that run before and after each of the lines above, e.g.

.. code:: python

   engine.run_event("before_forward")
   outputs = model.forward(batch)
   engine.run_event("after_forward")

Algorithms and callbacks (see below) register themselves to run on one
or more events.

We also maintain a :class:`.State` which stores the trainer's state, such as
the model, optimizers, dataloader, current batch, etc (see
:class:`.State`). This allows algorithms to modify the state at the
various events above.

.. seealso::

>     :doc:`Events<trainer/events>` and :class:`.State`

Reviewer: Events is not autolinking.

Algorithms
~~~~~~~~~~

The Composer trainer is designed to easily apply our library of
algorithms to both train more efficiently and build better models. These
can be enabled by passing the appropriate algorithm class to ``algorithms``
argument.

.. testcode::

   from composer import Trainer
   from composer.algorithms import LayerFreezing, Mixup

   trainer = Trainer(model=model,
                     train_dataloader=train_dataloader,
                     eval_dataloader=eval_dataloader,
                     max_duration='160ep',
                     algorithms=[
                         LayerFreezing(freeze_start=0.5, freeze_level=0.1),
                         Mixup(alpha=0.1),
                     ])

   # the algorithms will automatically be applied during the appropriate
   # points of the training loop
   trainer.fit()

We handle inserting those algorithms into the training loop and in the
right order.

.. seealso::

    Our :doc:`Algorithms<algorithms>` guide, and the individual
    :doc:`Method Cards </method_cards/methods_overview>` for each algorithm.

Optimizers & Schedulers

> -----------------------

Reviewer: Make it the same heading level...

```suggestion
~~~~~~~~~~~~~~~~~~~~~~~
```

You can easily specify which optimizer and learning rate scheduler to
use during training. Composer provides a library of various optimizers
and schedulers, but you can also include one of your own.

> .. code:: python

Reviewer: Does `..code` run the example via doctest?

Author: it does not -- these need to be converted to doctest

   from composer import Trainer
   from composer.models import ComposerResNet
   from composer.optim.scheduler import cosine_annealing_scheduler
   from torch.optim import SGD

   model = ComposerResNet(model_name="resnet50", num_classes=1000)

>    optimizer = SGD(model.parameters(), lr=0.1)

Reviewer: This example should be updated to use the new scheduling abstraction.

Author: I think showing that it supports the pytorch native schedulers is OK here.

   trainer = Trainer(model=model,
                     train_dataloader=train_dataloader,
                     eval_dataloader=eval_dataloader,
                     max_duration='90ep',
                     optimizer=optimizer,
                     scheduler=consine_annealing_scheduler)

.. seealso::

>     :doc:`Schedulers`<trainer/optimizers_and_schedulers>

Reviewer: Link is broken.

Training on GPU
~~~~~~~~~~~~~~~

Control which device you use for training with the ``device`` parameter,
and we will handle the data movement and other systems-related
engineering. We currently support the ``cpu`` and ``gpu`` devices.

.. code:: python

   from composer import Trainer

   trainer = Trainer(model=model,
                     train_dataloader=train_dataloader,
                     eval_dataloader=eval_dataloader,
                     max_duration='160ep',
                     device='gpu')

Distributed Training
~~~~~~~~~~~~~~~~~~~~

It's also simple to do data-parallel training on multiple GPUs. Composer
provides a launcher command that works with the trainer and handles all
the ``torch.distributed`` setup for you.

> 

Reviewer: Should we comment about the need for a `DistributedSampler` in the dataloader?

Reviewer: In addition, should we mention that, unlike grad accum, the dataloader batch size should not be scaled by the number of devices?

.. code:: python

   # run_trainer.py

   from composer import Trainer

   trainer = Trainer(model=model,
                     train_dataloader=train_dataloader,
                     eval_dataloader=eval_dataloader,
                     max_duration='160ep',

>                                        device='gpu')

Reviewer: indentration

   trainer.fit()

Access the Composer launcher via the ``composer`` command along with the
number of GPUs you'd like to use and your training script. Use
``composer --help`` to see a full list of configurable options.

.. code:: bash

   # run training on 8 GPUs

>    >>> composer -n 8 run_trainer.py

Reviewer: ```suggestion
   $ composer -n 8 run_trainer.py
```

.. seealso::

>     Our :doc:`Distributed Training<trainer/distributed_training>` guide and

Reviewer: link isn't linking

    the :mod:`composer.utils.dist` module.

DeepSpeed Integration
~~~~~~~~~~~~~~~~~~~~~

Composer comes with DeepSpeed support, allowing you to leverage their
full set of features that makes it easier to train large models across
(1) any type of GPU and (2) multiple nodes. For more details on DeepSpeed,
see `their website <https://www.deepspeed.ai>`__.

To enable DeepSpeed, simply pass in a config as specified in the
DeepSpeed docs `here <https://www.deepspeed.ai/docs/config-json/>`__.

.. code:: python

   # run_trainer.py

   from composer import Trainer

   trainer = Trainer(model=model,
                     train_dataloader=train_dataloader,
                     eval_dataloader=eval_dataloader,
                     max_duration='160ep',
                     device='gpu',

>                      deepspeed_config={

Reviewer: Would it work to pass an empty dictionary into the deepspeed config? Or do fields need to be set?
This would be an api change, but should we allow `deepspeed_config=True` to denote using the default values?

Author: Seems like an empty dict should work, but I don't see that tested anywhere. We should add an issue for this.

Author: Confirmed to be valid, adding some text to the this section.

                         "train_batch_size": 2048,
                         "amp": {"enabled": True},
                     })

.. warning::

    The ``deepspeed_config`` must not conflict with any other parameters
    passed to the trainer.

Callbacks
~~~~~~~~~

You can insert arbitrary callbacks to be run at various points during
the training loop. The Composer library provides several useful
callbacks for things such as monitoring throughput and memory usage
during training, but you can also implement your own.

.. code:: python

   from composer import Trainer
   from composer.callbacks import SpeedMonitor

   # include a callback for tracking throughput/step during training
   trainer = Trainer(model=model,
                     train_dataloader=train_dataloader,
                     eval_dataloader=eval_dataloader,
                     max_duration='160ep',
                     device='gpu',
                     callbacks=[SpeedMonitor(window_size=100)])

.. seealso::

    The :doc:`Callbacks<trainer/callbacks>` guide and :mod:`composer.callbacks`.

Numerics
~~~~~~~~

> Using `mixed precision <https://arxiv.org/abs/1710.03740>`__ can speed up your training loop and only requires

Reviewer: Can we add a blurb about what mixed precision is, and/or link to the pytorch docs for it? The paper is a bit intense...

points during training and (2) load them back to resume training later.

.. code:: python

   from composer import Trainer

   ### Saving checkpoints
   trainer = Trainer(model=model,
                     train_dataloader=train_dataloader,
                     eval_dataloader=eval_dataloader,
                     max_duration='160ep',
                     device='gpu',
                     # Checkpointing params
                     save_folder: 'checkpoints',
                     save_interval: '1ep')

   # will save checkpoints to the 'checkpoints' folder every epoch
   trainer.fit()

.. code:: python

   from composer import Trainer

   ### Loading checkpoints
   trainer = Trainer(model=model,
                     train_dataloader=train_dataloader,
                     eval_dataloader=eval_dataloader,
                     max_duration='160ep',
                     device='gpu',
                     # Checkpointing params
                     load_path: 'path/to/checkpoint/mosaic_states.pt')

   # will load the trainer state (including model weights) from the
   # load_path before resuming training
   trainer.fit()

.. seealso::

>     The :doc:`Checkpointing<trainer/checkpointing>` guide.

Reviewer: Link is broken

This was just a quick tour of all the features within our trainer. Please see the other
guides and notebooks for more information.

> Annotated Trainer Loop

Reviewer: The training loop code should be annotated via docstrings. Can we refactor the training loop code for readability? Ideally each function in the training loop should be as long as the pseudocode here so it is readable.

Author: This is out of scope for this PR

----------------------

> Our :class:`.Trainer` code is meant to be easily readable and understood. In this section,

Reviewer: Ahh I love the irony....if the training class is meant to be easily readable and understood, then I don't think we should need a separate section called "annotated trainer loop".

Author: 😂 you got me.

we walk you through the logic flow of the training loop code, from :meth:`.Trainer.fit`

> down to the `backward()` call.

Reviewer: ```suggestion
down to the :meth:`~torch.tensor.backward` call.
```

In pseudocode, the trainer is organized as follows:

.. code:: python

    def fit(self):
        try:
            _train_loop()
        finally:  # clean up
            self.engine.close()

> The method `_train_loop()` sets up the training, loads any

Reviewer: ```suggestion
The method ``_train_loop()`` sets up the training, loads any
```

.. code:: python

    # pseudocode
    def _train_loop(self):

        # setup training
        # metrics, gradient scaling, etc.
        # if needed, load checkpoints

        while timer < max_duration:

            for batch in train_dataloader:

                # for grad accum, split the batch
                microbatches = split_batch(batch)

                """"
                Depending on the config, the _train_batch()
                is called with slightly different code.
                """
                if deepspeed_enabled:
                    loss = self._train_batch(microbatches)
                elif _use_closure():
                    """
                    Mixed precision and some optimizers
                    requiring passing a closure.
                    """"
                    loss = optimizer.step(
                        closure=self._train_batch(microbatches)
                    )
                else:
                    loss = self._train_batch(microbatches)
                    optimizer.step()

                if eval_on_batch:
                    eval()  # run validation

            if eval_on_epoch:
                eval()

> Remaining are two methods: `_train_batch` and `_train_batch_inner`.

Reviewer: ```suggestion
Remaining are two methods: ``_train_batch`` and ``_train_batch_inner``.
```

> Remaining are two methods: `_train_batch` and `_train_batch_inner`.

Reviewer: This explanation makes it clear that `train_batch` and `train_batch_inner` are poor function names. Can we refactor it into `get_distributed_context_manager` and `train_microbatch`?

Author: Out of scope for this PR (no code changes)

----------

Below is a quick tour of various elements with code snippets for your reference. See the more
detailed sections in the navigation menu.

Events & State
~~~~~~~~~~~~~~

The core principle of the Composer trainer is to make it easy to inject
custom logic to run at various points in the training loop. To do this,
we have events that run before and after each of the lines above, e.g.

.. code:: python

   engine.run_event("before_forward")
   outputs = model.forward(batch)
   engine.run_event("after_forward")

Algorithms and callbacks (see below) register themselves to run on one
or more events.

We also maintain a :class:`.State` which stores the trainer's state, such as
the model, optimizers, dataloader, current batch, etc (see
:class:`.State`). This allows algorithms to modify the state at the
various events above.

.. seealso::

    :doc:`Events<events>` and :class:`.State`

Algorithms
~~~~~~~~~~

The Composer trainer is designed to easily apply our library of
algorithms to both train more efficiently and build better models. These
can be enabled by passing the appropriate algorithm class to ``algorithms``
argument.

.. testcode::

   from composer import Trainer

>    from composer.algorithms import LayerFreezing, Mixup

Reviewer 2: ``
```suggestion
   from composer.algorithms import LayerFreezing, MixUp
```

   trainer = Trainer(model=model,
                     train_dataloader=train_dataloader,
                     eval_dataloader=eval_dataloader,
                     max_duration='160ep',
                     algorithms=[
                         LayerFreezing(freeze_start=0.5, freeze_level=0.1),

>                          Mixup(alpha=0.1),

Reviewer 2: ```suggestion
                         MixUp(alpha=0.1),
```

data parallel across 8 GPUs, the dataloader should have ``batch_size=256``.

.. warning::

    If using distributed training, your dataloader use the
    :mod:`torch.utils.data.distributed.DistributedSampler`.

.. seealso::

    Our :doc:`Distributed Training<distributed_training>` guide and
    the :mod:`composer.utils.dist` module.

DeepSpeed Integration
~~~~~~~~~~~~~~~~~~~~~

Composer comes with DeepSpeed support, allowing you to leverage their
full set of features that makes it easier to train large models across
(1) any type of GPU and (2) multiple nodes. For more details on DeepSpeed,
see `their website <https://www.deepspeed.ai>`__.

To enable DeepSpeed, simply pass in a config as specified in the
DeepSpeed docs `here <https://www.deepspeed.ai/docs/config-json/>`__.

.. code:: python

   # run_trainer.py

   from composer import Trainer

   trainer = Trainer(model=model,
                     train_dataloader=train_dataloader,
                     eval_dataloader=eval_dataloader,
                     max_duration='160ep',
                     device='gpu',
                     deepspeed_config={
                         "train_batch_size": 2048,

>                          "amp": {"enabled": True},

Reviewer 2: Note: According to @jbloxham "amp" doesn't really work this way as documented by DeepSpeed. It would be better to change the to 

```
"fp16": {"enabled": True}
```

Author: fixed!
